ios/Codeforces Watcher/Application/components/FeedbackController.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  FeedbackController.swift
+//  Codeforces Watcher
+//
+//  Created by Den Matyash on 5/6/20.
+//  Copyright © 2020 xorum.io. All rights reserved.
+//
+
+import Foundation
+import StoreKit
+
+class FeedbackController {
+
+    private var KEY_CURRENT_SHOWING_ITEM: String {
+        get { "key_current_showing_item" }
+    }
+
+    private var KEY_COUNT_SCREEN_OPENING: String {
+        get { "key_count_app_opening" }
+    }
+
+    private var KEY_START_TIME_SHOWING: String {
+        get { "key_first_time_shown" }
+    }
+
+    private var KEY_IS_NEEDED_AT_ALL: String {
+        get { "key_is_needed_to_show" }
+    }
+
+    private var KEY_IS_LOCKOUT_PERIOD: String {
+        get { "key_is_lockout_period" }
+    }
+
+    private var currentShowingItem: Int {
+        get { UserDefaults.standard.integer(forKey: KEY_CURRENT_SHOWING_ITEM) }
+        set(value) { UserDefaults.standard.set(value, forKey: KEY_CURRENT_SHOWING_ITEM) }
+    }
+
+    private var countScreenOpening: Int {
+        get { UserDefaults.standard.integer(forKey: KEY_COUNT_SCREEN_OPENING) }
+        set(value) { UserDefaults.standard.set(value, forKey: KEY_COUNT_SCREEN_OPENING) }
+    }
+
+    private var startTimeWhenShown: Int64 {
+        get { UserDefaults.standard.object(forKey: KEY_START_TIME_SHOWING) as? Int64 ?? 0 }
+        set(value) { UserDefaults.standard.set(value, forKey: KEY_START_TIME_SHOWING) }
+    }
+
+    private var isNeededAtAll: Bool {
+        get { UserDefaults.standard.object(forKey: KEY_IS_NEEDED_AT_ALL) as? Bool ?? true }
+        set(value) { UserDefaults.standard.set(value, forKey: KEY_IS_NEEDED_AT_ALL) }
+    }
+
+    private var isLockoutPeriod: Bool {
+        get { UserDefaults.standard.bool(forKey: KEY_IS_LOCKOUT_PERIOD) }
+        set(value) { UserDefaults.standard.set(value, forKey: KEY_IS_LOCKOUT_PERIOD) }
+    }
+
+    var feedbackItem: FeedbackItem {
+        get {
+            switch currentShowingItem {
+            case 0: return buildFirst()
+            case 1: isNeededAtAll = false; return buildSecond()
+            case 2: return buildThird()
+            default:
+                fatalError()
+            }
+        }
+    }
+
+    private func buildFirst() -> FeedbackItem {
+        FeedbackItem(
+            textPositiveButton: "yes".localized,
```

Comment:
Here you can use just enum and then map strings in iOS / Android modules.

Reply:
could you elaborate?

Reply:
When extracted to `common` you can do the following `class FeedbackItem (val type: Type { enum Type { Feedback, App Store, Email }  }`. Then in Android `when (feedbackItem.type) {Feedback -> { bind strings to UI here }}, etc.`